packages/doenetml-worker-rust/lib-doenetml-core/src/dep_graph/prop_graph.rs

```diff
@@ -0,0 +1,171 @@
+use std::{cell::RefCell, rc::Rc};
+
+use crate::{
+    components::{ComponentEnum, ComponentNode},
+    Extending,
+};
+
+use super::graph::{Graph, Taggable};
+
+pub struct Components<'a> {
+    components: &'a Vec<Rc<RefCell<ComponentEnum>>>,
+}
+impl<'a> Components<'a> {
+    pub fn new(components: &'a Vec<Rc<RefCell<ComponentEnum>>>) -> Self {
+        Self { components }
+    }
+    /// Returns the index of the origin of this component:
+    ///  - If the component is extending another component, find the terminal referent.
+    ///  - Otherwise, return the index of the component itself.
+    pub fn origin(&self, component_idx: usize) -> usize {
+        match self.components[component_idx].borrow().get_extending() {
+            Some(&Extending::Component(source_idx)) => self.origin(source_idx),
+            _ => component_idx,
+        }
+    }
+}
+
+pub struct PropGraph {
+    pub graph: Graph<PropPointer, PropLookup<usize>>,
+}
+impl Default for PropGraph {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl PropGraph {
+    pub fn new() -> Self {
+        Self {
+            graph: Graph::new(),
+        }
+    }
+    //    pub fn resolve_prop(
+    //        &mut self,
+    //        prop_ptr: PropPointer,
+    //        components: &mut Vec<Rc<RefCell<ComponentEnum>>>,
+    //    ) {
+    //        let mut component = components[prop_ptr.component_idx].borrow_mut();
+    //        let mut prop = component.get_prop_mut(prop_ptr.prop_idx).unwrap();
+    //
+    //        let data_queries = prop.return_data_queries();
+    //        for query in data_queries {}
+    //        let components = Components::new(components);
+    //    }
+    //    pub fn add_nodes_from_query(
+    //        &mut self,
+    //        prop_ptr: PropPointer,
+    //        query: DataQuery,
+    //        components: &Components,
+    //    ) {
+    //        match query {
+    //            DataQuery::State => {
+    //                // State is a leaf node, but we if the component is extending another component
+    //                // we only want to create a single node on the graph relating to the original parent.
+    //                let origin = components.origin(prop_ptr.component_idx);
+    //                self.graph.add_node(PropPointer::new(origin, prop_ptr.prop_idx));
+    //            },
+    //            _ => panic!("Not implemented"),
+    //        }
+    //    }
+}
+
+/// Pointer to a prop on a specific component.
+#[derive(PartialEq, Eq, Clone, Copy, Hash)]
+pub struct PropPointer {
```

Comment:
Interesting abstraction of how it looks like you are hashing `PropPointer` but you're instead doing something more efficient. I just added a comment `graph.rs` about this, but it's not clear to me why `PropPointer` has to implement `Hash`.